lib/src/main/java/motif/ScopeFactory.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018-2019 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package motif;
+
+import motif.internal.Constants;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ScopeFactory {
+
+    private ScopeFactory() {}
+
+    private static final NoDependencies NO_DEPENDENCIES = new NoDependencies() {};
+
+    private static final Map<Class<?>, Constructor<?>> scopeClassToCreateMethod = new HashMap<>();
+
+    public static <S extends Creatable<NoDependencies>> S create(Class<S> scopeClass) {
+        return create(scopeClass, NO_DEPENDENCIES);
```

Comment:
And if someone sends in a `Creatable` with a non-zero arg constructor to this method? We would be passing `NO_DEPENDENCIES` to that constructor. That's legal, but incorrect usage. Do we want to warn or fail against that case?

Reply:
The class that we end up instantiating is the ScopeImpl associated with the Scope class that is passed in. If someone passes in a Creatable that is not a Scope then we will fail to find the ScopeImpl and crash at runtime. We could add an explicit error if the passed class is not annotated with `@Scope` but that adds some runtime overhead. Might be worth it - we can discuss this as a potential follow up.